ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/ReadyTransactions.java

```diff
@@ -139,30 +141,51 @@ public Stream<PendingTransaction> stream() {
   }
 
   @Override
-  public PendingTransaction promote(final Predicate<PendingTransaction> promotionFilter) {
-
-    final var maybePromotedTx =
-        orderByMaxFee.descendingSet().stream()
-            .filter(candidateTx -> promotionFilter.test(candidateTx))
-            .findFirst();
-
-    return maybePromotedTx
-        .map(
-            promotedTx -> {
-              final var senderTxs = txsBySender.get(promotedTx.getSender());
-              // we always promote the first tx of a sender, so remove the first entry
-              senderTxs.pollFirstEntry();
-              processRemove(senderTxs, promotedTx.getTransaction(), PROMOTED);
-
-              // now that we have space, promote from the next layer
-              promoteTransactions();
-
-              if (senderTxs.isEmpty()) {
-                txsBySender.remove(promotedTx.getSender());
-              }
-              return promotedTx;
-            })
-        .orElse(null);
+  public List<PendingTransaction> promote(
+      final Predicate<PendingTransaction> promotionFilter,
+      final long freeSpace,
+      final int freeSlots) {
+    long accSpace = 0;
+    final List<PendingTransaction> promotedTxs = new ArrayList<>();
+
+    // first find all txs that can be promoted
+    search:
+    for (final var senderFirstTx : orderByMaxFee.descendingSet()) {
+      final var senderTxs = txsBySender.get(senderFirstTx.getSender());
+      for (final var candidateTx : senderTxs.values()) {
+        if (promotionFilter.test(candidateTx)) {
+          accSpace += candidateTx.memorySize();
+          if (promotedTxs.size() < freeSlots && accSpace <= freeSpace) {
```

Comment:
maybe a comment to explain why < for slots and <= for space

Reply:
can I ask what you get from the code about the difference? to understand what needs clarification

Reply:
freeSlots and freeSpace have similar names - how much is free - so why does the if statement check < for one and <= for the other?

why not <= for freeSlots?

Reply:
slots always increment by 1, while for space it could be any amount, so you can read it like that

`if (promotedTxs.size() + 1 <= freeSlots && accumulatedSpace <= freeSpace)`

but the `+1` is redundant knowing that `promotedTxs.size()` can only increment by 1 at time, and we can shortcut with `<`

Reply:
right so it's because the accumulatedSpace is already incremented at this point but for the slots it happens after. makes sense. it's ok for me :)